source/SH3/camera/camera.cpp

```diff
@@ -0,0 +1,96 @@
+/** @file
+ *
+ *  Implementation of camera.hpp
+ *
+ *  @copyright 2017 Palm Studios
+ *
+ *  @date 1-9-2017
+ *
+ *  @author Jesse Buhagiar
+ */
+#include "SH3/camera/camera.hpp"
+#include "glm/gtc/matrix_transform.hpp"
+
+static constexpr glm::vec3 camFront = glm::vec3(0.0f, 0.0f, -1.0f); /**< Camera Front. Z is INTO the screen. */
+static constexpr glm::vec3 camUp    = glm::vec3(0, 1, 0);           /**< Camera Up. Y is UP. */
+
+sh3::camera::Camera::Camera(glm::vec3 pos, float fov, float aspect, float zNear, float zFar)
+{
+    camPos  = pos;
+    camFov  = fov;
+    aRatio  = aspect;
+    camNear = zNear;
+    camFar  = zFar;
+
+    pMatrix = glm::perspective(glm::radians(fov), aspect, zNear, zFar); // Set our perspective matrix
+    vMatrix = glm::lookAt(camPos, camPos + camFront, camUp);
+}
+
+void sh3::camera::Camera::RecalculatePerspective()
+{
+    vMatrix = glm::lookAt(camPos, camPos + camFront, camUp);
+    pMatrix = glm::perspective(glm::radians(camFov), aRatio, camNear, camFar); // Set our perspective matrix
+}
+
+glm::mat4 sh3::camera::Camera::GetProjectionMatrix() const
+{
+    return pMatrix * glm::lookAt(camPos, camPos + camFront, camUp);
+    //return pMatrix * vMatrix;
+}
+
+void sh3::camera::Camera::SetPosition(float x, float y, float z)
+{
+    camPos = glm::vec3(x, y, z);
+}
+
+void sh3::camera::Camera::SetPosition(glm::vec3 pos)
+{
+    camPos = pos;
+}
+
+void sh3::camera::Camera::Translate(float x, float y, float z)
+{
+    // This is _really_ fucking messy... Please make me less hacky!!! Thanks!
```

Comment:
Since `camFront` and `camUp` are always `(0, 0, -1)` and `(0, 1, 0)`, `camFront x camUp` is always `(1, 0, 0)`, correct? As you can see you don't even need to normalize, as both operands to the product are unit vectors, so the resulting vector is as well.

Now with a positive x, camPos get's translated by `-(x, 0, 0) = (-x, 0, 0)`.
With a negative x it's `(-x, 0, 0)`, which is identical.
Why do we need the different branches then? Even an `x = 0` is handled gracefully.

Reply:
Interesting, I completely neglected to realise this...